### DISCUSSION TITLE
37. CORS 요청 방식 세 가지를 설명해주세요.

### DISCUSSION THREAD

**Discussion by linirini** [jewelry-box/cs-interview#38] [General]
- 단순 요청에서도 서버가 origin 헤더를 확인하나요?
- 확인하는 주체가 서버일까요, 브라우저일까요?
- 단순 요청 예시에 대해서 아시나요?
- Credential request를 사용했을 때 CORS를 검사하는 추가되는 룰 2가지가 무엇인가요?

**Comment by jongmee** [comment]
### 단순 요청에서도 서버가 origin 헤더를 확인하나요?
단순 요청은 확인하지 않습니다.

### 확인하는 주체가 서버일까요, 브라우저일까요?
origin 헤더를 확인하는 건 서버입니다. 브라우저는 사전 요청에서 서버의 응답 헤더를 확인합니다.

### 단순 요청 예시에 대해서 아시나요?
대표적으로 HTML form element를 통한 요청이 있습니다.

### Credential request를 사용했을 때 CORS를 검사하는 추가되는 룰 2가지가 무엇인가요?
우선 Credential request는 쿠키가 포함된 요청, Authorization 헤더를 가진 요청을 말한다. 서버에서 Access-Control-Allow-Credentials를 true로 설정하고 origin, method, header의 범위에 와일드카드를 사용해서는 안된다.

**Reply by cutehumanS2** [reply]
단순 요청에서도 서버가 클라이언트가 보낸 Origin 헤더를 확인하지 않나요 ?!

**Comment by cutehumanS2** [comment]
## CORS 요청 방식 3가지
프리플라이트 요청, 단순 요청, 자격 증명 요청

### [프리플라이트 요청]
- 프리플라이트 요청 방식은 브라우저가 요청을 한번에 보내지 않고 예비 요청과 본 요청으로 나눠 보내는 방식으로, 여기서 예비 요청을 프리플라이트라고 합니다.
- **본 요청을 보내기 전에 요청의 유효성을 검사하기 위해** 예비 요청을 보내며, 이 과정에서 OPTIONS 메서드가 사용됩니다.
    - **서버가 반환하는 응답 헤더 종류:** 허용된 메서드, 헤더, 출처
        - Access-Control-Allow-Methods: 허용된 메서드
        - Access-Control-Allow-Headers: 허용된 헤더
        - Access-Control-Allow-Origin: 허용된 출처
    - 허용되지 않은 요청은 프리플라이트 단계에서 차단되므로, 보안 위협을 줄일 수 있습니다.

<br>

### [단순 요청 방식]
- 단순 요청 방식은 예비 요청을 보내지 않고, 바로 서버에게 본 요청을 보낸 후, 서버가 이에 대한 응답으로 헤더에 허용되는 출처 정보를 보내주면, 그때 브라우저가 CORS 정책 위반 여부를 검사하는 방식입니다.
- 예비 요청 유무를 제외하고는 프리플라이트와 전반적인 로직이 같습니다.
- 이 방식은 사용할 수 있는 조건이 까다롭기 때문에 거의 사용되지 않습니다.
    - 요청 메서드가 GET, HEAD, POST 중 하나여야 합니다.
    - Accept, Content-Type 등 특정 헤더만 사용해야 합니다.
    - 만약 Content-Type을 사용하는 경우, 특정 콘텐츠 타입만 허용됩니다: 자주 사용하는 applicaiton/json을 사용할 수 없습니다.

<br>

### [자격 증명 요청 방식]
- 요청에 쿠키나 토큰 같은 인증 관련 정보를 담을 수 있게 해주는 credentials 옵션을 통해 인증된 요청을 사용하는 방법입니다.
- credentials 옵션 3가지

    옵션 값 | 설명
    -- | --
    same-origin(기본값) | 같은 출처 간 요청에만 인증 정보를 담을 수 있다.
    include | 모든 요청에 인증 정보를 담을 수 있다.
    omit | 모든 요청에 인증 정보를 담지 않는다.

- 이 방식을 사용하면 CORS 정책 위반 여부를 검사하는 룰에 두 가지가 추가됩니다.
    - Access-Control-Allow-Origin 헤더에 와일드 카드를 사용할 수 없으며 명시적인 URL을 지정해야 합니다.
    - 응답 헤더에 Access-Control-Allow-Credentials: true 값이 반드시 포함되어야 합니다.
